angel_system/global_step_prediction/global_step_predictor.py

```diff
@@ -0,0 +1,596 @@
+                  "threshold_frame_count_weak" number of frames.
+                  When this happens, the tracker can "skip" a step.
+            "step_to_activity_id": [0,1,2,3,24,4,...]
+                - array type.
+                - steps_to_activities[current_step] is the current activity index
+                  the tracker is on. (step index != activity index).
+            "step_to_activity_desc": ["background", ...]
+            "starting_frame_index": 30
+                - index videos frame at which point this tracker instance was initialized.
+            "prediction_history": [0,0,0,...,1,1,1,...23,23]
+            "active": True
+                - When true, keep checking to update this recipe.
+                - When false, the recipe is complete... stop adding to its history
+                  or checking for a next step.
+
+        """
+        tracker_dict = {}
+        if recipe == "coffee":
+            config_fn = "config/tasks/task_steps_cofig-recipe-coffee-shortstrings.yaml"
+        else:
+            raise ValueError(f"Invalid recipe type. Valid types: [coffee].")
+        with open(config_fn, "r") as stream:
+            config = yaml.safe_load(stream)
+        labels = [self.sanitize_str(l["description"]) for l in config["steps"]]
+        steps = config["steps"]
+        if steps[0]["id"] == 1:
+            config["steps"].insert(
+                0,
+                {
+                    "id": 0,
+                    "activity_id": 0,
+                    "description": "background",
+                    "median_duration_seconds": 0.5,
+                    "mean_conf": 0.5,
+                    "std_conf": 0.2,
+                },
+            )
+        tracker_dict["recipe"] = recipe
+        tracker_dict["current_step"] = 0
+        tracker_dict["total_num_steps"] = len(steps)
+        tracker_dict["can_skip"] = False
+        tracker_dict["step_to_activity_id"] = [step["activity_id"] for step in steps]
+        tracker_dict["step_to_activity_desc"] = [step["description"] for step in steps]
+        tracker_dict["prediction_history"] = np.array([])
+        tracker_dict["active"] = True
+        tracker_dict["steps"] = steps
+
+        self.trackers.append(tracker_dict)
+
+    def process_new_confidences(self, activity_confs):
+        assert np.array(activity_confs).shape[1] <= len(self.activated_activities)
+
+        activated_indexes = np.where(self.activated_activities[:, 0] == 1)[0]
+        deactivated_indexes = np.where(self.activated_activities[:, 0] == 0)[0]
+        assert len(activated_indexes) + len(deactivated_indexes) == len(
+            self.activated_activities
+        )
+
+        # activity conf = vector of all activities for one frame.
+        for i, activity_conf in enumerate(activity_confs):
+            if i == 6000:
+                self.initialize_new_recipe_tracker("coffee")
+                print(f"initialized second tracker")
```

Comment:
This seems pretty hard-coded. What is this? If this should stick around it needs some contextualizing commentary.